module/zfs/vdev_raidz_math.c

```diff
@@ -0,0 +1,626 @@
+}
+
+static raidz_rec_f
+_reconstruct_fun_raidz3(raidz_map_t *rm, const int *parity_valid,
+	const int nbaddata)
+{
+	if (nbaddata == 1) {
+		if (parity_valid[CODE_P]) {
+			return (rm->rm_ops->rec[RAIDZ_REC_P]);
+		} else if (parity_valid[CODE_Q]) {
+			return (rm->rm_ops->rec[RAIDZ_REC_Q]);
+		} else if (parity_valid[CODE_R]) {
+			return (rm->rm_ops->rec[RAIDZ_REC_R]);
+		}
+	} else if (nbaddata == 2) {
+		if (parity_valid[CODE_P] && parity_valid[CODE_Q]) {
+			return (rm->rm_ops->rec[RAIDZ_REC_PQ]);
+		} else if (parity_valid[CODE_P] && parity_valid[CODE_R]) {
+			return (rm->rm_ops->rec[RAIDZ_REC_PR]);
+		} else if (parity_valid[CODE_Q] && parity_valid[CODE_R]) {
+			return (rm->rm_ops->rec[RAIDZ_REC_QR]);
+		}
+	} else if (nbaddata == 3 &&
+		parity_valid[CODE_P] && parity_valid[CODE_Q] &&
+		parity_valid[CODE_R]) {
+		return (rm->rm_ops->rec[RAIDZ_REC_PQR]);
+	}
+	return ((raidz_rec_f) NULL);
+}
+
+/*
+ * Select data reconstruction method for raidz_map
+ * @parity_valid - Parity validity flag
+ * @dt           - Failed data index array
+ * @nbaddata     - Number of failed data columns
+ */
+int
+vdev_raidz_math_reconstruct(raidz_map_t *rm, const int *parity_valid,
+	const int *dt, const int nbaddata)
+{
+	raidz_rec_f rec_data = NULL;
+
+	switch (raidz_parity(rm)) {
+		case 1:
+			rec_data = _reconstruct_fun_raidz1(rm, parity_valid,
+			    nbaddata);
+			break;
+		case 2:
+			rec_data = _reconstruct_fun_raidz2(rm, parity_valid,
+			    nbaddata);
+			break;
+		case 3:
+			rec_data = _reconstruct_fun_raidz3(rm, parity_valid,
+			    nbaddata);
+			break;
+		default:
+			cmn_err(CE_PANIC, "invalid RAID-Z configuration %d",
+			    raidz_parity(rm));
+			break;
+	}
+
+	ASSERT(rec_data != NULL);
+
+	return (rec_data(rm, dt));
+}
+
+
+
+static void
+init_raidz_kstat(raidz_math_kstat_t *rs, const char *name)
+{
+	int i;
+	const size_t impl_name_len = strnlen(name, KSTAT_STRLEN);
+	const size_t op_name_max = (KSTAT_STRLEN - 2) > impl_name_len ?
+		KSTAT_STRLEN - impl_name_len - 2 : 0;
+
+	for (i = 0; i < RAIDZ_GEN_NUM; i++) {
+		strncpy(rs->gen[i].name, name, impl_name_len);
+		strncpy(rs->gen[i].name + impl_name_len, "_", 1);
+		strncpy(rs->gen[i].name + impl_name_len + 1,
+			raidz_gen_name[i], op_name_max);
+
+		rs->gen[i].data_type = KSTAT_DATA_UINT64;
+		rs->gen[i].value.ui64  = 0;
+	}
+
+	for (i = 0; i < RAIDZ_REC_NUM; i++) {
+		strncpy(rs->rec[i].name, name, impl_name_len);
+		strncpy(rs->rec[i].name + impl_name_len, "_", 1);
+		strncpy(rs->rec[i].name + impl_name_len + 1,
+			raidz_rec_name[i], op_name_max);
+
+		rs->rec[i].data_type = KSTAT_DATA_UINT64;
+		rs->rec[i].value.ui64  = 0;
+	}
+}
+
+#define	BENCH_D_COLS	(8ULL)
+#define	BENCH_COLS	(BENCH_D_COLS + PARITY_PQR)
+#define	BENCH_COL_SIZE	(1ULL << 12)
+#define	BENCH_ZIO_SIZE	(BENCH_COL_SIZE * BENCH_D_COLS)
+#define	BENCH_NS	MSEC2NSEC(25)			/* 25ms */
+
+typedef void (*benchmark_fn)(raidz_map_t *rm, const int fn);
+
+static void
+__benchmark_gen(raidz_map_t *rm, const int fn)
```

Comment:
Same as above private `__*` functions should be renamed `*_impl` for consistency.